R/learn_cvpairs.R

```diff
@@ -1,37 +1,37 @@
 #' Cross-validation pairs
 #'
-#' Add cross validation pairs to a pipelearner object. \code{n} random
-#' partitions will try to be created unless \code{k} is not NULL, in which case
-#' k folds will be created. See \code{\link[modelr]{crossv_mc}} for more
-#' details.
+#' Add cross validation pairs to a pipelearner object. By default will create
+#' a random partition of the data. Can use any function which creates
+#' \code{train} and \code{test} columns of \code{resample} objects and column
+#' of \code{id} integers. See \code{\link[modelr]{crossv_mc}} and
+#' \code{\link[modelr]{resample}} for details.
 #'
 #' Will expect a pipelearner object, but will also accept a data frame for the
 #' parameter \code{pl}. In the case that a data frame is used, it will first
 #' coerce it to a pipelearner object via \code{\link{pipelearner}}.
 #'
 #' @inheritParams pipelearner_params
-#' @inheritParams modelr::crossv_mc
+#' @param .f a function which takes a \code{data} argument
+#' @param ... additional parameters to be passed to \code{.f}
+#'
+#' @examples
+#' pl <- pipelearner(mtcars)
+#' learn_cvpairs(pl, modelr::crossv_kfold, k = 4)
+#' learn_cvpairs(pl, resamplr::holdout_frac, size = 0.3, K = 5L)
 #' @export
-learn_cvpairs <- function(pl, k = NULL, n = 1, test = 0.2) {
+learn_cvpairs <- function(pl, .f = function(data) modelr::crossv_mc(data, n = 1), ...) {
   UseMethod("learn_cvpairs")
 }
 
 #' @export
-learn_cvpairs.default <-  function(pl, k = NULL, n = 1, test = 0.2) {
-  learn_cvpairs.pipelearner(pipelearner(pl), k = k, n = n, test = test)
+learn_cvpairs.default <-  function(pl, .f = function(data) modelr::crossv_mc(data, n = 1), ...) {
+  learn_cvpairs.pipelearner(pipelearner(pl), ...)
 }
 
 #' @export
-learn_cvpairs.pipelearner <- function(pl, k = NULL, n = 1, test = 0.2) {
-
-  # n random partitions
-  if (is.null(k)) {
-    cv_pairs <- modelr::crossv_mc(pl$data, n = n, test = test, id = ".id")
-  } else {
-    cv_pairs <- modelr::crossv_kfold(pl$data, k = k, id = ".id")
-  }
-
-  pl$cv_pairs <- cv_pairs
+learn_cvpairs.pipelearner <- function(pl, .f = function(data) modelr::crossv_mc(data, n = 1), ...) {
+  pl$cv_pairs <- purrr::invoke(.f, data = pl$data, ...)
```

Comment:
Can cope with default situation here via `if (missing(.f)) {...}`. Must be sure to carefully document this default case.